File: src/gestorAplicación/clasesPrincipales/Factura.java
package gestorAplicación.clasesPrincipales;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Iterator;

import gestorAplicación.clasesHeredadas.Cliente;

public class Factura implements Serializable {
	private Reserva reserva;
    private String facturaHecha;
    private Cliente cliente;
	private Mesa mesa;
	private int precio;
	private String sede;
	private String id;
	private Membresia membresia;

	
    public Factura(Reserva reserva) {
    	this.reserva = reserva;
    	this.cliente = reserva.getCliente();
		this.mesa = reserva.getMesa();
		this.sede = mesa.getSede();
		this.id = reserva.getIdR();
		this.membresia = cliente.getMembresia();
		this.precio = 100000;
		Pago.addFacturasPendientes(this);
    }
    

    public Reserva getReserva() {
    	return reserva;
    }
    public void setReserva(Reserva reserva) {
    	this.reserva = reserva;
    }

	public Cliente getClienteFactura(){
		return reserva.getCliente();
	}

	public Mesa getMesa() {
    	return mesa;
    }

	public String getFacturaHecha() {
		return facturaHecha;
		}
	
	public int getPrecio() {
		return precio;
	}
	public void setPrecio(int precio) {
		this.precio = precio;
	}
	
	public String getSede() {
		return sede;
	}    

	public Membresia getFacturaMiembro() {
		return membresia;
	}
	
	public String getIDReserva() {
	    return id;
	}


	private boolean clienteEsMiembro(Cliente cliente) {
		Iterator<Cliente> iterator = Cliente.getMiembrosActuales().iterator();
		while (iterator.hasNext()) {
			Cliente clienteActual = iterator.next();
			if (clienteActual.getNombre().equals(cliente.getNombre()) && clienteActual.getId() == cliente.getId()) {
				return true; 
			}
		}
		return false; 
		
	}
	
public String toString() {
    String facturaHecha = "-------------\nRestaurante Un\nId de la reserva: " + id + "\nCliente: " + cliente.getNombre() + "\nIdentificación: " + cliente.getId()
            + "\nSede: " + sede + "\nMesa #" + mesa.getId() + "\nHora: " + reserva.getHora();

    if (clienteEsMiembro(cliente)) {
        facturaHecha += "\nEl cliente es miembro Gold";
        int precioConDescuento = precio - 20000;
        facturaHecha += "\nDescuento del 20% aplicado\nPrecio: 80000$";
        if (precio == 100000) {
            this.precio = precioConDescuento;
        }
    } else {
        facturaHecha += "\nEl cliente no es miembro";
        facturaHecha += "\nPrecio: " + precio + "$";
    }

    facturaHecha += "\n-------------";
    return facturaHecha;
}


	public static Factura buscarFactura(ArrayList<Factura> facturas,String id){
        for (Factura f : facturas){
			if (f.getReserva().getIdR().trim().equals(id)){
				return f;
			}
		}
		return null;
    }
	public void actualizarFactura(Reserva reserva){
		Pago.removePendiente(this);
		new Factura(reserva);
	}
}

File: Python/src/gestorAplicación/clasesPrincipales/Factura.py
from gestorAplicación.clasesPrincipales import Pago
from gestorAplicación.clasesHeredadas import Cliente

class Factura:

    def __init__(self, reserva):
        self._reserva = reserva
        self._cliente = reserva.getCliente()
        self._mesa = reserva.getMesa()
        self._sede = self._mesa.getSede()
        self._id = reserva.getIdR()
        self._membresia = self._cliente.getMembresia()
        self._precio = 100000    
        Pago.addFacturasPendientes(self)
    
    #Métodos get
    def getReserva(self):
        return self._reserva
    
    def getClienteFactura(self):
        return self._cliente
    
    def getMesa(self):
        return self._mesa
    
    #def getFacturaHecha(self):
    #    return self._facturaHecha
    
    def getPrecio(self):
        return self._precio
    
    def getSede(self):
        return self._sede
    
    def getFacturaMiembro(self):
        return self._membresia
    
    def getIDReserva(self):
        return self._id
    
    #Métodos set
    def setReserva(self, reserva):
        self._reserva = reserva
    
    def setPrecio(self, precio):
        self._precio = precio

    #Otros métodos
    @staticmethod
    def buscarFactura(facturas, id):
        for factura in facturas:
            if factura.getReserva().getIdR().strip() == id:
                return factura
        return None

    def clienteEsMiembro(self, cliente):
        lista = Cliente.getMiembrosActuales()
        for clienteActual in lista:
            if (clienteActual.getNombre() == cliente.getNombre) and (clienteActual.getId() == cliente.getId):
                return True
        return False
    
    def actualizarFactura(self, reserva):
        Pago.removePendiente(self)
        Factura(reserva)

    #ToString
    def __str__(self):
        facturaHecha = "-------------\nRestaurante Un\nId de la reserva: {}\nCliente: {}\nIdentificación: {}\nSede: {}\nMesa #{}\nHora: {}" 
        facturaHecha = facturaHecha.format(self._id, self._cliente.getNombre(), self._cliente.getId(), self._sede, self._mesa.getId(), self._reserva.getHora())

        if self.clienteEsMiembro(self._cliente):
            facturaHecha + "\nEl cliente es miembro Gold"
            precioConDescuento = self.precio - 20000
            facturaHecha += "\nDescuento del 20% aplicado\nPrecio: 80000$"
            if self._precio == 100000:
                self.precio = precioConDescuento             
        else:
            facturaHecha += "\nEl cliente no es miembro"
            facturaHecha += "\nPrecio: {}$".format(self._precio)
        
        facturaHecha += "\n-------------"
        return facturaHecha